api/buckets/service.go

```diff
@@ -132,86 +177,119 @@ func (s *Service) Create(ctx context.Context, req *pb.CreateRequest) (*pb.Create
 		return nil, ErrTooManyBucketsInThread
 	}
 
-	var key []byte
-	if req.Private {
-		var err error
-		key, err = dcrypto.NewKey()
-		if err != nil {
-			return nil, err
-		}
-	}
 	var bootCid cid.Cid
 	if req.BootstrapCid != "" {
 		bootCid, err = cid.Decode(req.BootstrapCid)
 		if err != nil {
 			return nil, fmt.Errorf("invalid bootstrap cid: %s", err)
 		}
 	}
-	buck, seed, err := s.createBucket(ctx, dbID, dbToken, req.Name, key, bootCid)
+	buck, seed, err := s.createBucket(ctx, dbID, dbToken, req.Name, req.Private, bootCid)
 	if err != nil {
 		return nil, err
 	}
 	var seedData []byte
-	if key != nil {
-		seedData, err = decryptData(seed.RawData(), key)
+	if buck.IsPrivate() {
+		fileKey, err := buck.GetFileEncryptionKeyForPath("")
+		if err != nil {
+			return nil, err
+		}
+		seedData, err = decryptData(seed.RawData(), fileKey)
 		if err != nil {
 			return nil, err
 		}
 	} else {
 		seedData = seed.RawData()
 	}
 
+	root, err := getPbRoot(dbID, buck)
+	if err != nil {
+		return nil, err
+	}
 	return &pb.CreateResponse{
-		Root: &pb.Root{
-			Key:       buck.Key,
-			Name:      buck.Name,
-			Path:      buck.Path,
-			Thread:    dbID.String(),
-			CreatedAt: buck.CreatedAt,
-			UpdatedAt: buck.UpdatedAt,
-		},
+		Root:    root,
 		Links:   s.createLinks(dbID, buck),
 		Seed:    seedData,
 		SeedCid: seed.Cid().String(),
 	}, nil
 }
 
 // createBucket returns a new bucket and seed node.
-func (s *Service) createBucket(ctx context.Context, dbID thread.ID, dbToken thread.Token, name string, key []byte, bootCid cid.Cid) (buck *tdb.Bucket, seed ipld.Node, err error) {
+func (s *Service) createBucket(ctx context.Context, dbID thread.ID, dbToken thread.Token, name string, private bool, bootCid cid.Cid) (buck *tdb.Bucket, seed ipld.Node, err error) {
+	var owner thread.PubKey
+	if dbToken.Defined() {
+		owner, err = dbToken.PubKey()
+		if err != nil {
+			return nil, nil, fmt.Errorf("creating bucket: invalid token public key")
+		}
+	} else {
+		owner = getOwnerFromContext(ctx)
+	}
+
+	// Create bucket keys if private
+	var linkKey, fileKey []byte
+	if private {
```

Comment:
This is creating the single link key and the initial top-level file key. More file keys would be added if paths were opened to other users.